plugins/woocommerce/includes/abstracts/abstract-wc-settings-api.php

```diff
@@ -327,6 +327,10 @@ public function generate_settings_html( $form_fields = array(), $echo = true ) {
 
 			if ( method_exists( $this, 'generate_' . $type . '_html' ) ) {
 				$html .= $this->{'generate_' . $type . '_html'}( $k, $v );
+			} elseif ( has_action( 'woocommerce_generate_' . $type . '_html' ) ) {
+				ob_start();
+				do_action( 'woocommerce_generate_' . $type . '_html', $k, $v, $this );
+				$html .= ob_get_clean();
```

Comment:
Sorry for the delay in getting to this PR.

Would you be able to change this to a filter (ie, `$html .= apply_filters( ... )`)? The existing approach certainly works, but we feel the use of an action is generally best reserved for signalling an event took place. A filter would also remove the need for output buffering, and so keep the code a little shorter.

Last but not least, it would be great if a [docblock](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#4-hooks-actions-and-filters) could be added to accompany the hook.

Reply:
In converting the action into a filter, I had two possible choices: pass `$html` and let the callback function deal with the string concatenation; OR pass an empty string and concatenate the result with `$html` once the filter returns.

I decided to go with this second option because each iteration should not be allowed, albeit by accident, to alter the markup generated in the previous iterations. So, I believe what you see in this new commit is the safest approach that satisfies your request.

I hope this works for you.